### PR TITLE
Remove old vacancies

### PR DESCRIPTION
**Pull request by g4s8**
#971 - remove old vacancies

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#1255](https://github.com/zerocracy/farm/issues/1255) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#1255](https://github.com/zerocracy/farm/issues/1255) is assigned to @krzyk[/z](https://www.0crat.com/u/krzyk), here is [why](https://www.0crat.com/footprint/C3NDPUA8L/1771369568); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/main/groovy/com/zerocracy/stk/pmo/vacancies/remove_old_vacancies.groovy [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import com.zerocracy.pmo.Vacancies
+
+import java.time.LocalDateTime
+import java.time.ZoneId
+
+def exec(Project pmo, XML xml) {
+  new Assume(pmo, xml).isPmo()
+  new Assume(pmo, xml).type('Ping daily')
+  ClaimIn claim = new ClaimIn(xml)
+  LocalDateTime timestamp = LocalDateTime.ofInstant(claim.created().toInstant(), ZoneId.systemDefault())
```
Comment (krzyk) [inline src/main/groovy/com/zerocracy/stk/pmo/vacancies/remove_old_vacancies.groovy]: @g4s8 wouldn't it be better to operate on UTC times and use `ZonedDatetime`?

File: src/main/groovy/com/zerocracy/stk/pmo/vacancies/remove_old_vacancies.groovy [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  Farm farm = binding.variables.farm
```
Comment (krzyk) [inline src/main/groovy/com/zerocracy/stk/pmo/vacancies/remove_old_vacancies.groovy]: @g4s8 `farm` can be inlined

File: src/main/java/com/zerocracy/pmo/Vacancies.java [hunk 3]
```diff
@@ -81,14 +83,28 @@ public Vacancies bootstrap() throws IOException {
...
+     * Add new vacancy.
+     * @param project Hiring project
+     * @param author Vacancy author
+     * @param text Vacancy text
+     * @param added Added time
+     * @throws IOException If fails
+     * @checkstyle ParameterNumberCheck (4 lines)
+     */
+    public void add(final Project project, final String author,
+        final String text, final LocalDateTime added) throws IOException {
```
Comment (krzyk) [inline src/main/java/com/zerocracy/pmo/Vacancies.java]: @g4s8 `ZonedDateTime` with UTC timezone has this advantage that the time is known precisely without added burden of day light saving

File: src/test/java/com/zerocracy/pmo/VacanciesTest.java [hunk 4]
```diff
@@ -53,4 +57,40 @@ public void removesVacancy() throws Exception {
...
+        );
+        vacancies.removeOlderThan(
+            LocalDateTime.of(
+                LocalDate.of(2018, Month.APRIL, 1),
+                LocalTime.of(0, 0)
+            )
+        );
+        MatcherAssert.assertThat(
+            vacancies.iterate(),
+            Matchers.contains(Matchers.equalTo(second.pid()))
```
Comment (krzyk) [inline src/test/java/com/zerocracy/pmo/VacanciesTest.java]: @g4s8 `Matchers.equalTo` shouldn't be needed

File: src/test/resources/com/zerocracy/bundles/remove_old_vacancies/_after.groovy [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+import com.jcabi.xml.XML
+import com.zerocracy.Farm
+import com.zerocracy.Project
+import com.zerocracy.pmo.Vacancies
+import org.hamcrest.MatcherAssert
+import org.hamcrest.Matchers
+
+def exec(Project project, XML xml) {
+  Farm farm = binding.variables.farm
```
Comment (krzyk) [inline src/test/resources/com/zerocracy/bundles/remove_old_vacancies/_after.groovy]: @g4s8 you can inline this

**Comment by g4s8** [comment]
@krzyk fixed, please check. Also, look at this comment about `farm` inlining: https://github.com/zerocracy/farm/pull/1254#discussion_r198175529

**Comment by krzyk** [comment]
@g4s8 yeah, but the inlining works for me. I mean I don't see any errors in the build or tests, and I don't need to do any cast. So what is the problem with it?

**Comment by krzyk** [comment]
@rultor mege pls

**Comment by g4s8** [comment]
@rultor mergre

**Comment by g4s8** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @krzyk[/z](https://www.0crat.com/u/krzyk), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#1255](https://github.com/zerocracy/farm/issues/1255) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by ypshenychka** [comment]
@0crat quality good

**Comment by 0crat** [comment]
> @0crat quality good ([here](https://github.com/zerocracy/farm/issues/1255#issuecomment-400429544))

@ypshenychka There is no quality review for [#1255](https://github.com/zerocracy/farm/issues/1255), no performer

**Comment by 0crat** [comment]
Order was finished, quality is "good": +20 point(s) just awarded to @krzyk[/z](https://www.0crat.com/u/krzyk)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Comment by g4s8** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/zerocracy/farm/issues/1255#issuecomment-400635012))

@g4s8 This is what I know about this job in [C3NDPUA8L](https://www.0crat.com/p/C3NDPUA8L), as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#1255](https://github.com/zerocracy/farm/issues/1255) is not in scope
  * Job [footprint](https://www.0crat.com/footprint/C3NDPUA8L?q=%7Bjob%3A%27gh%3Azerocracy%2Ffarm%231255%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Pull request merged**